File: cpp/linear.cpp
#include "linear.hpp"

namespace linear {
    Eigen::VectorXi get_order(const Eigen::MatrixXi& A) {
        int n = A.rows();
        int m = A.cols();

        // Sum along rows to find zeros
        Eigen::VectorXi weights(n);
        for (int i = 0; i < n; ++i) {
            if (A.row(i).sum() == 0) {
                weights(i) = m;
            }
            else {
                weights(i) = argmax(get_binary(A.row(i)));
            }
        }
        return argsort(weights);
    }

    std::pair<Eigen::MatrixXi, Eigen::MatrixXi> Gauss_elimination(const Eigen::MatrixXi& input_A, const Eigen::MatrixXi& input_b, int N, bool allow_reorder) {
        Eigen::MatrixXi A = input_A;
        Eigen::MatrixXi b = input_b;

        int n = A.rows();
        int m = A.cols();

        if (n == 0 || m == 0) {
            return std::make_pair(A, b);
        }

        Eigen::VectorXi inv_list = get_inv_list(N);

        auto div = [&N, &inv_list](int a, int b) { return (a * inv_list(b)) % N; };

        auto update = [&div, &N, &n](Eigen::MatrixXi& old_A, Eigen::MatrixXi& old_b, int k, int pivot, std::pair<int, int> slic) {
            auto [begin, end] = slic;
            Eigen::MatrixXi new_A = old_A;
            Eigen::MatrixXi new_b = old_b;
            for (int i = begin; i < end; ++i) {
                int f = div(old_A(i, pivot), old_A(k, pivot));
                new_A.row(i) = mod((old_A.row(i) - old_A.row(k) * f).array(), N);
                new_b.row(i) = mod((old_b.row(i) - old_b.row(k) * f).array(), N);
            }
            return std::make_pair(new_A, new_b);
        };

        std::vector<int> pivots;
        for (int k = 0; k < n; ++k) {
            int pivot = argmax(get_binary(A.row(k)));
            pivots.push_back(pivot);
            auto slic = std::make_pair(k+1, n);
            std::tie(A, b) = update(A, b, k, pivot, slic);
        }

        for (int k = n - 1; k >= 0; --k) {
            int pivot = pivots[k];
            auto slic = std::make_pair(0, k);
            std::tie(A, b) = update(A, b, k, pivot, slic);

            int inv = inv_list(A(k, pivot));
            if (inv != 0) {
                A.row(k) = mod((A.row(k) * inv).array(), N);
                b.row(k) = mod((b.row(k) * inv).array(), N);
            }
        }

        if (allow_reorder) {
            Eigen::VectorXi order = get_order(A);
            A = A(order, Eigen::all).eval();
            b = b(order, Eigen::all).eval();
        }

        return std::make_pair(A, b);
    }

    std::tuple<Eigen::VectorXb, bool, Eigen::MatrixXi> solve_modN(const Eigen::MatrixXi& A, const Eigen::MatrixXi& b, int N, bool canonicalized) {

        int n = A.rows();
        int m = A.cols();
        int k = b.cols();

        assert(b.rows() == n);

        if (m == 0) {
            Eigen::VectorXb is_over = (b.array() > 0).colwise().any();
            Eigen::MatrixXi x = Eigen::MatrixXi::Zero(m, k);
            return std::make_tuple(is_over, false, x);
        }

        Eigen::MatrixXi A2, b2;
        if (canonicalized) {
            A2 = A;
            b2 = b;
        } else {
            std::tie(A2, b2) = Gauss_elimination(A, b, N, true);
        }

        Eigen::VectorXi conditions = A2.rowwise().sum();
        Eigen::MatrixXb cond1 = (conditions.array() == 0).rowwise().replicate(k);
        Eigen::MatrixXb cond2 = (b2.array() != 0);
        Eigen::VectorXb is_over_ref = (cond1 && cond2).colwise().any();

        Eigen::MatrixXi b2_zero = b2(get_where(conditions.array() == 0), Eigen::all);
        Eigen::VectorXb is_over = (b2_zero.array() > 0).colwise().any();

        assert((is_over_ref.array() == is_over.array()).all());

        bool is_under = (conditions.array() > 1).any();

        Eigen::VectorXb nonzero = (conditions.array() > 0);
        Eigen::VectorXi nonzero_indices = get_where(nonzero);
        A2 = A2(nonzero_indices, Eigen::all).eval();
        b2 = b2(nonzero_indices, Eigen::all).eval();

        auto pivots = map_rowwise_scalar<int>(argmax, A2);
        Eigen::MatrixXi x = Eigen::MatrixXi::Zero(m, k);
        x(pivots, Eigen::all) = b2;

        return std::make_tuple(is_over, is_under, x);
    }

    Eigen::MatrixXi Gauss_elimination_Aonly(const Eigen::MatrixXi& A, int N, bool allow_reorder) {
        Eigen::MatrixXi b = Eigen::MatrixXi::Zero(A.rows(), 0);
        auto [A2, b2] = Gauss_elimination(A, b, N, allow_reorder);
        return A2;
    }

    std::tuple<Eigen::MatrixXi, Eigen::MatrixXi> Gauss_elimination_full(const Eigen::MatrixXi& A, int N, bool allow_reorder) {
        int n = A.rows();
        Eigen::MatrixXi b = Eigen::MatrixXi::Identity(n, n);
        return Gauss_elimination(A, b, N, allow_reorder);
    }

    Eigen::MatrixXi get_kernel_reordered(const Eigen::MatrixXi& raw_A2, int N) {
        Eigen::VectorXb nonzero = raw_A2.array().rowwise().sum() > 0;
        Eigen::MatrixXi A2 = raw_A2(get_where(nonzero), Eigen::all).eval();
        int n = A2.rows();
        int m = A2.cols();

        if (m == 0) {
            return Eigen::MatrixXi::Zero(0, 0);
        }

        Eigen::VectorXi main_cols = map_rowwise_scalar<int>(argmax, get_binary(A2));
        Eigen::VectorXi tmp_array = arange(m, m * 2);
        tmp_array(main_cols) = arange(0, n);
        //set_values<int>(tmp_array, main_cols, arange(0, n));

        Eigen::VectorXi col_order = argsort(tmp_array);
        //Eigen::MatrixXi A2_sort = get_cols(A2, col_order);
        Eigen::MatrixXi A2_sort = A2(Eigen::all, col_order);

        Eigen::MatrixXi b_sort = Eigen::MatrixXi::Identity(m - n, m - n);
        Eigen::MatrixXi a_sort = mod(-A2_sort.rightCols(m - n).array(), N);

        Eigen::MatrixXi x_sort(m, m - n);
        x_sort << a_sort, b_sort;
        Eigen::MatrixXi x = Eigen::MatrixXi::Zero(m, m - n);
        //set_rows_values(x, col_order, x_sort);
        x(col_order, Eigen::all) = x_sort;
        return x;
    }

    Eigen::MatrixXi get_kernel(const Eigen::MatrixXi& A, int N) {
        Eigen::MatrixXi A2 = Gauss_elimination_Aonly(A, N, true);
        return get_kernel_reordered(A2, N);
    }

    Eigen::VectorXi simplify_coset_representative(const Eigen::MatrixXi& A2, const Eigen::VectorXi& x, int N) {
        if (A2.rows() == 0) {
            return x;
        }

        Eigen::VectorXi arg = map_rowwise_scalar<int>(argmax, get_binary(A2));
        // assert((A2.rowwise().sparseView().value(arg) == 1).all());
        //return mod(x - A2.transpose() * get_values(x, arg), N);
        return mod((x - A2.transpose() * x(arg)).array(), N);
    }
}

File: py/linear.py
import numpy as np


class UnderDetermined(Exception):
    pass


class OverDetermined(Exception):
    pass


def get_inv_list(N, dtype=None):
    x = np.arange(N, dtype=dtype)
    return np.argmax((x[:, None] * x) % N == 1, axis=1)


def get_order(A):
    n, m = A.shape[-2:]
    zero = np.sum(A, axis=-1) == 0
    order = np.argsort(np.argmax(A > 0, axis=-1) + zero * m, axis=-1, kind='stable')
    return order


def Gauss_elimination(A, b, N, allow_reorder=False, verbose=False):
    # convert A by row transformation to
    # I C
    # O O
    if np.prod(A.shape) == 0:
        return A.copy(), b.copy()
    A = A.copy()
    b = b.copy()
    inv_list = get_inv_list(N, dtype=A.dtype)

    n, m = A.shape[-2:]
    shape = A.shape[0:-2]
    indices = tuple(np.indices(shape))

    def div(a, b):
        return (a * inv_list[b]) % N

    def update(A, b, k, pivot, slic):
        f = div(A[indices + (slic, pivot)], A[indices + (slice(k, k+1), pivot)])
        A[..., slic, :] = (A[..., slic, :] - A[..., k:k+1, :] * f[..., :, None]) % N
        b[..., slic, :] = (b[..., slic, :] - b[..., k:k+1, :] * f[..., :, None]) % N

    if verbose:
        print(np.concatenate((A, b), axis=-1))
        print()

    pivots = []
    pivot = 0
    for k in range(n):
        pivot = np.argmax(A[..., k, :] > 0, axis=-1)  # ..., 0-m
        slic = slice(k+1, n)
        update(A, b, k, pivot, slic)
        if verbose:
            print(np.concatenate((A, b), axis=-1))
            print()
        pivots.append(pivot)

    for k in range(n - 1, -1, -1):
        pivot = pivots[k]  # ..., 0-m
        slic = slice(0, k)
        update(A, b, k, pivot, slic)
        inv = inv_list[A[indices + (k, pivot)]]

        do_update = (inv[..., None] != 0)
        A_k_update = (A[..., k, :] * inv[..., None]) % N
        b_k_update = (b[..., k, :] * inv[..., None]) % N
        A[..., k, :] = A_k_update * do_update + A[..., k, :] * np.logical_not(do_update)
        b[..., k, :] = b_k_update * do_update + b[..., k, :] * np.logical_not(do_update)
        if verbose:
            print(np.concatenate((A, b), axis=-1))
            print()
    if allow_reorder:
        indices_new = tuple([item[..., None] for item in indices])
        order = get_order(A)
        A = A[indices_new + (order, )]
        b = b[indices_new + (order, )]
    return A, b


def Gauss_elimination_Aonly(A, N, allow_reorder=False, verbose=False):
    n, m = A.shape[-2:]
    b = np.zeros(A.shape[0:-2] + (n, 0), dtype=A.dtype)
    return Gauss_elimination(A, b, N, allow_reorder=allow_reorder, verbose=verbose)[0]


def Gauss_elimination_full(A, N, allow_reorder=False, verbose=False):
    n, m = A.shape[-2:]
    b = np.zeros(A.shape[0:-2] + (n, n), dtype=A.dtype)
    arange = np.arange(n)
    b[..., arange, arange] = 1
    return Gauss_elimination(A, b, N, allow_reorder=allow_reorder, verbose=verbose)


def get_col_order(arg, N):
    # step1: array = [4, 5, 6, 7]
    # step2: array = [4, 0, 6, 1]
    # step3: order = [1, 3, 0, 2]
    shape = arg.shape[0:-1]
    n = arg.shape[-1]
    indices = tuple(np.indices(shape))
    indices_new = tuple([item[..., None] for item in indices])

    array = np.broadcast_to(np.arange(N, N * 2), shape + (N, )).copy()
    array[indices_new + (arg, )] = np.broadcast_to(np.arange(n), shape + (n, ))
    return np.argsort(array, axis=-1)


def get_kernel_reordered(A2, N):
    # assume A2 is obtained by Gaussian_elimination(xxx, allow_reorder=True)
    A2 = A2[np.sum(A2, axis=1) > 0]
    n, m = A2.shape[-2:]
    if m == 0:
        return np.zeros((0, 0), dtype=int)
    assert n <= m
    shape = A2.shape[0:-2]
    indices = tuple(np.indices(shape))
    indices_new = tuple([item[..., None] for item in indices])

    assert np.all(np.any(A2, axis=-1))

    col_order = get_col_order(np.argmax(A2 > 0, axis=-1), m)
    # A2_sort: reorder columns of A so that A2_sort = (I, C), I: nxn, C: nx(m-n)
    # x = (a, b)^T, a: nx(m-n), b: (m-n)x(m-n), x: mx(m-n)
    # b = I, a = -Cb = -C
    A2_sort = A2[indices_new + (slice(None), col_order)]
    b_sort = np.broadcast_to(np.eye(m - n, dtype=A2.dtype), shape + (m - n, m - n))
    a_sort = (- A2_sort[..., :, n:]) % N
    x_sort = np.concatenate((a_sort, b_sort), axis=-2)  # (..., m, m - n)
    x = np.zeros_like(x_sort)
    x[indices_new + (col_order, slice(None))] = x_sort
    return x


def get_kernel(A, N):
    # linear space: {x|Ax=0 (mod N)}
    A2 = Gauss_elimination_Aonly(A, N, allow_reorder=True)
    return get_kernel_reordered(A2, N)


def get_kernel_reordered_bool(A2):
    return get_kernel_reordered(A2.astype(int), 2).astype(bool)


def get_kernel_bool(A):
    return get_kernel(A.astype(int), 2).astype(bool)


def solve_modN(A, b, N, verbose=False):
    if b.ndim == 1:
        is_over, is_under, x = solve_modN(A, b[:, None], N)
        return is_over[0], is_under, x[:, 0]
    n, m = A.shape
    assert b.shape[0] == n
    k = b.shape[1]

    if m == 0:
        is_over = np.any(b > 0, axis=0)
        is_under = False
        x = np.zeros((m, k), dtype=int)
        return is_over, is_under, x

    A2, b2 = Gauss_elimination(A, b, N, verbose=verbose)
    conditions = np.sum(A2, axis=1)
    is_over = np.any(b2[conditions == 0] > 0, axis=0)
    is_under = np.any(conditions > 1)
    A2 = A2[conditions > 0]
    b2 = b2[conditions > 0]
    pivots = np.argmax(A2, axis=1)
    x = np.zeros((m, k), dtype=int)
    x[pivots] = b2.copy()
    return is_over, is_under, x


def simplify_coset_representative(A2, x, N):
    if np.prod(A2.shape) == 0:
        return x
    argmax = np.argmax(A2 > 0, axis=-1)
    assert np.all(A2[np.arange(len(A2)), argmax] == 1)
    return (x - x[argmax].dot(A2)) % N


if __name__ == '__main__':
    N = 3
    n = 4
    A = np.random.randint(0, N, size=(5, n, n))
    b = np.zeros_like(A)
    b[:, np.arange(n), np.arange(n)] = 1
    A2, b2 = Gauss_elimination(A, b, N)
    assert np.all(np.einsum("xij,xjk->xik", b2, A) % N == A2)
